### PR TITLE
fix(pip): Add the ability to handle pip's --extra-index-url

### LINKED ISSUE #4035: [python] Add support for --extra-index-url
In our current setup, we don't want to enforce the registry to be used, and we usually specify an extra index url where the packages can be found.

This can be found in the `requirements.txt` and `requirements.in` files

Example `requirements.txt`:

```
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --output-file requirements.txt requirements.in
#
--extra-index-url https://pypi.example.com/pypi

arrow==0.12.1
attrs==17.4.0
certifi==2018.1.18        # via requests
cffi==1.11.4              # via pynacl
chardet==3.0.4            # via requests
click==6.7                # via flask
```

Generated from some `requirements.in` file like:

```
structlog
raven
redis
requests
webargs

--extra-index-url https://pypi.example.com/pypi
our-private-package
```

### PR DESCRIPTION
**Pull request by RmStorm**
As specified in the docs here:
https://pip.pypa.io/en/stable/reference/pip_wheel/#extra-index-url
it's possible to define more index-urls in a single package by using
--extra-index-url this was not correctly handled by renovatebot.

Makes renovatebot correctly recognize --extra-index-url in pip requirements.txt files.

Closes #4035

### CODE REVIEW

**Comment by rarkins** [comment]
@RmStorm thanks for the PR!

Can you verify that the following use case is handled?
- `--index-url` is *not* specified
- One or more `--extra-index-url` is specified

Based on previous discussions, I was under the impression that the default `pypi` registry should be tried first in such situation, but I think it will be missing from `registryUrls` in the current approach.

**Comment by RmStorm** [comment]
> @RmStorm thanks for the PR!
> 
> Can you verify that the following use case is handled?
> 
>     * `--index-url` is _not_ specified
> 
>     * One or more `--extra-index-url` is specified
> 
> 
> Based on previous discussions, I was under the impression that the default `pypi` registry should be tried first in such situation, but I think it will be missing from `registryUrls` in the current approach.

Hi, so in the case that there is no --index-url specified do you think it would be best to simply add 'https://pypi.org/pypi/' to registryUrls in addition to the extra index? Or should it try to fetch the default that is specified in renovatebots configuration?

I'm not sure what the best approach is. I think adding pypi makes sense. But that is overwritable in pip config ourisde of repositories if I'm correct. I guess in that case people would to like to get the default and the --extra-index-url but I'm not sure if that's how it works.

**Comment by rarkins** [comment]
I'm not an expert on it but that was the conclusion we reached here: https://github.com/renovatebot/renovate/issues/4035#issuecomment-509670014

**Comment by rarkins** [comment]
i.e. if index-url is unspecified but extra-index-url is, then the user expects that the default pypi index is still used.

**Comment by RmStorm** [comment]
Hmm, so am I correct in thinking that if a default repo is supplied to renovatebot in someway it is stored in process.env.PIP_INDEX_URL? Because in that case --index-url is currently ignored anyways in the case a different default is supplied [or so it seems](https://github.com/renovatebot/renovate/blob/cfaa33423660a5e4f7880933078c22513af94bc7/lib/datasource/pypi/index.js#L35)?

Can you explain me how an overwritten default is handled/passed by renovatebot? I am obviously unfamiliar with the codebase, sorry for that.

**Comment by rarkins** [comment]
`process.env.PIP_INDEX_URL` is only relevant to self-hosted users. It's not used in the app.

**Comment by rarkins** [comment]
In the app you need to configure via `registryUrls` using `packageRules`

**Comment by RmStorm** [comment]
> In the app you need to configure via `registryUrls` using `packageRules`

Hmm, ok. But how is it used in the code? I assume after it's configured it should show up in [this function](https://github.com/renovatebot/renovate/blob/cfaa33423660a5e4f7880933078c22513af94bc7/lib/datasource/pypi/index.js#L30) somehow and overwrite the default [host_url here](https://github.com/renovatebot/renovate/blob/cfaa33423660a5e4f7880933078c22513af94bc7/lib/datasource/pypi/index.js#L31) right?

**Comment by RmStorm** [comment]
I'm having a lot of trouble figuring out where the `registryUrls` created in [this function](https://github.com/renovatebot/renovate/blob/3d04c696fba297da91d860153f0c3b178c129fee/lib/manager/pip_requirements/extract.js#L19) is combined with the `registryUrls` defined in the config [here](https://github.com/renovatebot/renovate/blob/3d04c696fba297da91d860153f0c3b178c129fee/lib/config/definitions.js#L412).

Is it combined? or does one simply take precedence at some point before `registryUrls` is used [here](https://github.com/renovatebot/renovate/blob/cfaa33423660a5e4f7880933078c22513af94bc7/lib/datasource/pypi/index.js#L30)? I don't think I can spend a lot more time on this but I would really like to wrap this up.

**Comment by rarkins** [comment]
They’re empty by default. If left empty then the pypi datasource will use the public pypi only. 

If custom URLs are to be used then they need to all be specified in registryUrls. If no custom URLs are necessary then it’s recommended not to set them during the extract function.

**Comment by rarkins** [comment]
in that pypi datasource any non empty array takes precedence over the default registry.

**Comment by rarkins** [comment]
In short: it needs to follow the logic mentioned in the issue this Pr is closing. Needs awareness of whether index-url is set or not

**Comment by RmStorm** [comment]
> In short: it needs to follow the logic mentioned in the issue this Pr is closing. Needs awareness of whether index-url is set or not

Yes I understand. The issue was that it was quite difficult to figure out how to combine the registryUrls from the config with the new ones from the pip requirement file. Turns out that config is being passed into the extractPackageFile function. It was just not explicitly accepted in the case of the pip_requirements manager. I think it should be pretty good now. Let me know if something else needs to be changed before we can get this in.

**Comment by rarkins** [comment]
Thanks, looks good to the eye. I’ll merge once I’ve done some manual testing

File: lib/manager/pip_requirements/extract.js [hunk 1]
```diff
@@ -16,16 +16,31 @@ module.exports = {
...
+    if (line.startsWith('--extra-index-url ')) {
+      const registryUrl = line
+        .substring('--extra-index-url '.length)
+        .split(' ')[0];
+      registryUrls.push(registryUrl);
     }
   });
+  if (registryUrls.length > 0 && !has_index_url) {
+    if (config.registryUrls) {
+      registryUrls.push(...config.registryUrls);
```
Comment (rarkins) [inline lib/manager/pip_requirements/extract.js]:
```suggestion
      registryUrls.unshift(...config.registryUrls);
```

File: lib/manager/pip_requirements/extract.js [hunk 2]
```diff
@@ -16,16 +16,31 @@ module.exports = {
...
+    } else {
+      registryUrls.push('https://pypi.org/pypi/');
```
Comment (rarkins) [inline lib/manager/pip_requirements/extract.js]:
```suggestion
      registryUrls.unshift('https://pypi.org/pypi/');
```

File: lib/manager/pip_requirements/extract.js [hunk 3]
```diff
@@ -16,16 +16,31 @@ module.exports = {
...
   logger.trace('pip_requirements.extractPackageFile()');
 
-  let registryUrls;
+  const registryUrls = [];
+  let has_index_url = false;
   content.split('\n').forEach(line => {
     if (line.startsWith('--index-url ')) {
       const registryUrl = line.substring('--index-url '.length).split(' ')[0];
-      registryUrls = [registryUrl];
+      registryUrls.push(registryUrl);
```
Comment (rarkins) [inline lib/manager/pip_requirements/extract.js]: index-url should always be first in the array before extra-index-url. Do we need to worry about the case where the ordering is reversed in `requirements.txt`?

**Review by rarkins — CHANGES_REQUESTED** [review]
Pypi registries are checked sequentially in the order defined, so the "index url" needs to always be first:
- index-url should be put before extra-index-url in `registryUrls` even if they are not ordered that way in the requirements.txt (can that happen?)
- If no index url specified then the default URL should be put to the front of the array, not back

**Pull request merged**